### ISSUE TITLE
Dataset: numpy array instead of images?

### ISSUE THREAD

**Issue by GiorgioBarnabo** [junyanz/pytorch-CycleGAN-and-pix2pix#1092]
Hi everybody,

I was wondering whether I could feed the network with 256*256 NumPy array instead of jpeg images. I have been dealing with mel-spectrograms and it would be quite hard to turn them into images.

Thank you,

Giorgio

**Comment by junyanz** [comment]
It is possible. You need to modify the data loader. Here is a data loader [template](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/data/template_dataset.py).

**Comment by GiorgioBarnabo** [comment]
Thank you very much!

**Comment by GiorgioBarnabo** [comment]
Just one quick another thing. Can't I just change can change the following two lines:

        A_img = Image.open(A_path).convert('RGB')
        B_img = Image.open(B_path).convert('RGB')

Into:

        A_img = np.load(A_path)
        B_img = np.load(B_path)

Without modifying the custom data loader template?

**Comment by junyanz** [comment]
You may want to also modify the function `make_dataset`. See this [lline](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/data/unaligned_dataset.py#L29).

**Comment by GiorgioBarnabo** [comment]
Dear @junyanz ,

sorry for disturbing you again. I would need some more advice. I was finally able to train my model with the NumPy arrays, but something went wrong. I need to train an image2image model that takes as input 256*256 2D numpy arrays whose values go from -40 to +40 (they represent melspectrograms), and outputs numpy arrays with the same characteristics. 

Here is how I modified the data loader for the training (UnalignedDataset):

    def __init__(self, opt):
        """Initialize this dataset class.
 
        Parameters:
            opt (Option class) -- stores all the experiment flags; needs to be a subclass of BaseOptions
        """
        BaseDataset.__init__(self, opt)
        self.dir_A = os.path.join(opt.dataroot, opt.phase + 'A')  # create a path '/path/to/data/trainA'
        self.dir_B = os.path.join(opt.dataroot, opt.phase + 'B')  # create a path '/path/to/data/trainB'
 
        self.A_paths = sorted(make_dataset(self.dir_A, opt.max_dataset_size))   # load images from '/path/to/data/trainA'
        self.B_paths = sorted(make_dataset(self.dir_B, opt.max_dataset_size))    # load images from '/path/to/data/trainB'
        self.A_size = len(self.A_paths)  # get the size of dataset A
        self.B_size = len(self.B_paths)  # get the size of dataset B
        btoA = self.opt.direction == 'BtoA'
        input_nc = self.opt.output_nc if btoA else self.opt.input_nc       # get the number of channels of input image
        output_nc = self.opt.input_nc if btoA else self.opt.output_nc      # get the number of channels of output image
        self.transform_A = get_transform(self.opt, grayscale=(input_nc == 1))
        self.transform_B = get_transform(self.opt, grayscale=(output_nc == 1))
 
    def __getitem__(self, index):
        """Return a data point and its metadata information.
 
        Parameters:
            index (int)      -- a random integer for data indexing
 
        Returns a dictionary that contains A, B, A_paths and B_paths
            A (tensor)       -- an image in the input domain
            B (tensor)       -- its corresponding image in the target domain
            A_paths (str)    -- image paths
            B_paths (str)    -- image paths
        """
        A_path = self.A_paths[index % self.A_size]  # make sure index is within then range
        if self.opt.serial_batches:   # make sure index is within then range
            index_B = index % self.B_size
        else:   # randomize the index for domain B to avoid fixed pairs.
            index_B = random.randint(0, self.B_size - 1)
        B_path = self.B_paths[index_B]
        A_img = np.load(A_path)
        A_img = np.expand_dims(A_img, axis=0)
        B_img = np.load(B_path)
        B_img = np.expand_dims(B_img, axis=0)
        # apply image transformation
        A = torch.from_numpy(A_img)
        B = torch.from_numpy(B_img)
 
        return {'A': A, 'B': B, 'A_paths': A_path, 'B_paths': B_path}

And here's the one for the test:

class SingleDataset(BaseDataset):
    """This dataset class can load a set of images specified by the path --dataroot /path/to/data.

    It can be used for generating CycleGAN results only for one side with the model option '-model test'.
    """

    def __init__(self, opt):
        """Initialize this dataset class.

        Parameters:
            opt (Option class) -- stores all the experiment flags; needs to be a subclass of BaseOptions
        """
        BaseDataset.__init__(self, opt)
        self.A_paths = sorted(make_dataset(opt.dataroot, opt.max_dataset_size))
        input_nc = self.opt.output_nc if self.opt.direction == 'BtoA' else self.opt.input_nc
        self.transform = get_transform(opt, grayscale=(input_nc == 1))

    def __getitem__(self, index):
        """Return a data point and its metadata information.

        Parameters:
            index - - a random integer for data indexing

        Returns a dictionary that contains A and A_paths
            A(tensor) - - an image in one domain
            A_paths(str) - - the path of the image
        """
        A_path = self.A_paths[index]
        A_img = np.load(A_path)
        A_img = np.expand_dims(A_img, axis=0)
        A = torch.from_numpy(A_img)
        return {'A': A, 'A_paths': A_path}

I do not know why but after 7 epochs of training (36000 "images" each), the test output makes no sense: numpy array 256*256*3 with values ranging in a very large interval. 

Finally, here is how I launched the code:

python train.py --dataroot "/content/drive/Shared drives/MusicBERT_Datasets/CycleGAN_Dataset" --name voice2song --model cycle_gan --input_nc 1 --output_nc 1 --save_epoch_freq 1

python test.py --dataroot "/content/drive/Shared drives/MusicBERT_Datasets/CycleGAN_Dataset/testA" --name voice2song --model test --input_nc 1 --output_nc 1 --no_dropout

Any idea?

Thanks a lot!!

**Comment by junyanz** [comment]
I am not sure. You might want to normalize the data to [-1,1]. Our current generator only outputs values between [-1, 1]. For images, you can use `transform_A` and `transform_B`. But they may not work for your numpy array. You need to normalize the data by yourself. For debugging purpose, you can print the shape of your torch array as well as the min and max value.

**Comment by GiorgioBarnabo** [comment]
Thank you very much! The training eventually worked! Just another quick question: after training the model can I test it in both directions right?

**Comment by JadeCrystal0228** [comment]
@GiorgioBarnabo Hi, I also used MelSpectrogram (e.g., float32) as input to do the translation. As @junyanz mentioned earlier, the generator only outputs values between [-1, 1]. I was wondering how to convert the generated values back to melspectrogram, which is fed into the following CNN model. How did you do that?

**Comment by polewins** [comment]
> @GiorgioBarnabo Hi, I also used MelSpectrogram (e.g., float32) as input to do the translation. As @junyanz mentioned earlier, the generator only outputs values between [-1, 1]. I was wondering how to convert the generated values back to melspectrogram, which will be fed into the subsequent CNN model (e.g., a 2-class classification model). How did you do that?

Hi, did you ever manage to work out how to output your generated values back to melspectrogram? Stuck on what to edit to get my output values back as melspoectrograms rather than png
